### ISSUE TITLE
Clarification on Mathematical Derivation of Unknown Logit Term

### ISSUE THREAD

**Issue by 1345265244** [feifeiobama/OrthogonalDet#19]
Dear Author,

Thank you for sharing your valuable research work. It provides great insights and inspiration to the community.

Regarding the code snippet you mentioned:
```
class_logit_max = class_logits[:, :-1].max(dim=-1, keepdim=True).values
class_logits -= class_logit_max
tmp = class_logits[:, :-1].exp().sum(dim=-1, keepdim=True)
unknown_logit = (tmp - 1).log() + tmp.log() - (1 + class_logits[:, -1:].exp()).log()
```

I have some questions about its specific meaning and logic. Could you please briefly explain the design intention behind this code, especially the principle of calculating unknown_logitin the last line?

**Comment by feifeiobama** [comment]
this code is derived from equation 3 in the paper, specifically `p(unknown) = p_original(known - max_known)` as referred in [this implementation](https://github.com/feifeiobama/OrthogonalDet/blob/main/core/head.py#L253).

let the original class logits be denoted by $\{l_i\}_{i\in \mathcal{K}\cup\{u\}}$, and $l_m$ represent the max known logit. then the adjust unknown logit $l_u'$ is defined to satisfy:

$$
\frac{e^{l_u'}}{\sum_{i\in\mathcal{K}}e^{l_i}+e^{l_u'}}=\frac{\sum_{i\in\mathcal{K}}e^{l_i}-e^{l_m}}{\sum_{i\in\mathcal{K}}e^{l_i}+e^{l_u}}.
$$

from this, we obtain:

$$
\begin{aligned}
e^{l_u'}&=\frac{(\sum_{i\in\mathcal{K}}e^{l_i}-e^{l_m})\sum_{i\in\mathcal{K}}e^{l_i}}{e^{l_u}+e^{l_m}}\\
&=\frac{(\mathrm{tmp}-1)\mathrm{tmp}}{1+e^{l_u-l_m}}
\end{aligned},
$$

where $\mathrm{tmp}=\sum_{i\in\mathcal{K}}e^{l_i-l_m}$.

**Comment by 1345265244** [comment]
Thank you for your answer. It was very helpful.

**Comment by 1345265244** [comment]
@feifeiobama  Thank you so much for your help with my previous question. It was very clear. I was wondering if I could ask one more thing.

In the source code's config file, the mask parameter in t2.yaml is set to 2 in order to mask the logits of previously seen classes. However, after I commented out this parameter (using dual RTX 3090s) and evaluated only on T2 (without using T2_ft), the results showed that performance was only poorer on A-OSE, while other metrics actually seemed better than T2_FT.

Could you please explain the role of mask=2?

`[10/20 09:27:30] core.pascal_voc_evaluation INFO: Wilderness Impact: {0.1: {50: 0.005795235028976176}, 0.2: {50: 0.00822150822150822}, 0.3: {50: 0.00922202562865262}, 0.4: {50: 0.011098678440741526}, 0.5: {50: 0.01198716866452811}, 0.6: {50: 0.01245837505530587}, 0.7: {50: 0.013790084272737223}, 0.8: {50: 0.012929352645580846}, 0.9: {50: 0.010975353591933903}}
[10/20 09:27:32] core.pascal_voc_evaluation INFO: avg_precision: {0.1: {50: 0.013402204029437207}, 0.2: {50: 0.010776415821531672}, 0.3: {50: 0.009213095921883975}, 0.4: {50: 0.007501876784699469}, 0.5: {50: 0.006591008911746787}, 0.6: {50: 0.006591008911746787}, 0.7: {50: 0.006591008911746787}, 0.8: {50: 0.006591008911746787}, 0.9: {50: 0.006591008911746787}}
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Absolute OSE (total_num_unk_det_as_known): {50: 2749.0}
[10/20 09:27:32] core.pascal_voc_evaluation INFO: total_num_unk 10693
[10/20 09:27:32] core.pascal_voc_evaluation INFO: ['aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus', 'car', 'cat', 'chair', 'cow', 'diningtable', 'dog', 'horse', 'motorbike', 'person', 'pottedplant', 'sheep', 'sofa', 'train', 'tvmonitor', 'truck', 'traffic light', 'fire hydrant', 'stop sign', 'parking meter', 'bench', 'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'frisbee', 'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard', 'tennis racket', 'banana', 'apple', 'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'bed', 'toilet', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone', 'book', 'clock', 'vase', 'scissors', 'teddy bear', 'hair drier', 'toothbrush', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'unknown']
[10/20 09:27:32] core.pascal_voc_evaluation INFO: AP50: ['76.8', '62.5', '52.3', '26.9', '35.9', '74.2', '56.0', '83.6', '29.0', '68.1', '26.6', '72.7', '79.5', '74.6', '63.1', '26.7', '65.2', '49.9', '77.2', '56.5', '19.5', '33.5', '71.3', '62.0', '46.2', '21.3', '75.0', '70.0', '83.2', '85.5', '6.3', '35.5', '5.2', '19.2', '30.4', '34.3', '20.4', '2.4', '25.5', '40.8', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', 'nan', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', 'nan', '0.0', '0.0', '0.5']
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Precisions50: ['12.5', '10.6', '11.8', '4.6', '6.4', '32.1', '13.6', '24.2', '7.4', '23.2', '6.1', '27.6', '23.7', '16.5', '17.4', '10.7', '13.0', '15.4', '23.0', '18.4', '6.6', '7.3', '14.4', '16.9', '11.6', '2.6', '21.0', '8.6', '28.7', '38.7', '3.6', '7.2', '2.4', '4.4', '6.7', '6.2', '4.7', '2.8', '4.5', '5.6', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.7']
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Recall50: ['90.4', '81.4', '70.4', '73.7', '75.1', '83.9', '81.5', '95.3', '63.4', '79.7', '62.2', '88.8', '89.7', '87.8', '88.7', '68.0', '82.8', '80.0', '90.1', '81.5', '79.0', '67.2', '88.9', '74.0', '60.0', '54.5', '92.5', '94.4', '91.4', '92.7', '48.2', '71.4', '42.3', '56.5', '69.1', '79.2', '67.6', '22.2', '72.4', '76.6', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', 'nan', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', 'nan', '0.0', '0.0', '43.5']
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Prev class AP50: 57.8797211241011
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Prev class Precisions50: 15.919718512254406
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Prev class Recall50: 80.71745155507844
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Current class AP50: 39.37675941824362
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Current class Precisions50: 10.217190583935267
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Current class Recall50: 70.01077485162844
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Known AP50: 48.628240271172366
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Known Precisions50: 13.068454548094838
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Known Recall50: 75.36411320335344
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Unknown AP50: 0.5190385791462266
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Unknown Precisions50: 0.6590458002830554
[10/20 09:27:32] core.pascal_voc_evaluation INFO: Unknown Recall50: 43.505096792294026`

**Comment by feifeiobama** [comment]
i cannot recall the details, but it seems to be an experimental setting (to be consistent with baselines) instead of method choice.

**Comment by 1345265244** [comment]
> i cannot recall the details, but it seems to be an experimental setting (to be consistent with baselines) instead of method choice.

it seems that setting mask=2 is intended to mask the logits of seen classes when calculating the cross-entropy loss. However, in my experiments on M-OWODB, I'm seeing a bit of evidence that not masking the seen class logits might lead to better performance? I might be wrong, but that's what the initial results suggest. [https://github.com/feifeiobama/OrthogonalDet/blob/d36b8e546a5427a4c5485b4ad65ecb12b118b1ca/core/loss.py#L55](url)

**Comment by 1345265244** [comment]
@feifeiobama Hello,
I need to correct my previous report. I apologize for the oversight — I had overlooked the existing filtering logic for cfg.TEST.MASKin the annotation loading process:`# filter instances
if cfg.TEST.MASK == 1:
    allowed_class = list(range(0, cfg.TEST.PREV_INTRODUCED_CLS+cfg.TEST.CUR_INTRODUCED_CLS))
else:
    allowed_class = list(range(cfg.TEST.PREV_INTRODUCED_CLS, cfg.TEST.PREV_INTRODUCED_CLS+cfg.TEST.CUR_INTRODUCED_CLS))`
